File: python/open_guardrail/guards/api_endpoint_safety.py
"""Validate API endpoint references for safety."""
from __future__ import annotations

import re
import time
from typing import Optional

from open_guardrail.core import GuardResult

_DEFAULT_PATTERNS: list[re.Pattern[str]] = [
    re.compile(
        r"/(admin|debug|internal|_debug|_admin"
        r"|_internal)/",
        re.IGNORECASE,
    ),
    re.compile(
        r"/(api/v0|api/beta|api/test|api/dev)\b",
        re.IGNORECASE,
    ),
    re.compile(r"localhost:\d+/api", re.IGNORECASE),
    re.compile(r"127\.0\.0\.1:\d+/api", re.IGNORECASE),
    re.compile(
        r"/(swagger|graphql-playground"
        r"|phpmyadmin|actuator)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"/(health|metrics|status|info|env)"
        r"(?:/|\s|$)",
        re.IGNORECASE,
    ),
    re.compile(r"/(\.env|\.git|\.config|wp-admin)"),
    re.compile(r"/api/deprecated/", re.IGNORECASE),
]


class _ApiEndpointSafety:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "api-endpoint-safety"
        self.action = action
        self._patterns = list(_DEFAULT_PATTERNS)

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for pat in self._patterns:
            if pat.search(text):
                matched.append(pat.pattern)

        triggered = len(matched) > 0
        score = (
            min(len(matched) / 3, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="api-endpoint-safety",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Unsafe API endpoint detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"matched_patterns": len(matched)}
                if triggered
                else None
            ),
        )


def api_endpoint_safety(
    *, action: str = "block"
) -> _ApiEndpointSafety:
    return _ApiEndpointSafety(action=action)

File: packages/guards/src/api-endpoint-safety.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ApiEndpointSafetyOptions {
  action: 'block' | 'warn';
}

const PATTERNS: RegExp[] = [
  /\/(admin|debug|internal|_debug|_admin|_internal)\//i,
  /\/(api\/v0|api\/beta|api\/test|api\/dev)\b/i,
  /localhost:\d+\/api/i,
  /127\.0\.0\.1:\d+\/api/i,
  /\/api\/[^v\s][^/\s]*(?:\/[^/\s]+)*(?!\bv\d)/i,
  /\/(swagger|graphql-playground|phpmyadmin|actuator)\b/i,
  /\/api(?!\/v\d)[^/\s]*\s/i,
  /\/(health|metrics|status|info|env)(?:\/|\s|$)/i,
  /\/(\.env|\.git|\.config|wp-admin)/i,
  /\/api\/deprecated\//i,
];

export function apiEndpointSafety(options: ApiEndpointSafetyOptions): Guard {
  return {
    name: 'api-endpoint-safety',
    version: '0.1.0',
    description: 'Validates API endpoint references for safety',
    category: 'security',
    supportedStages: ['input'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of PATTERNS) {
        if (pattern.test(text)) {
          matched.push(pattern.source);
        }
      }

      const triggered = matched.length > 0;
      const score = triggered ? Math.min(matched.length / 3, 1.0) : 0;

      return {
        guardName: 'api-endpoint-safety',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedPatterns: matched.length } : undefined,
      };
    },
  };
}
